### PR TITLE
Improve TexlFunctionSet

### PR DESCRIPTION
**Pull request by LucGenetier**
config.EnableJsonFunctions was generating a full copy of all dictionaries storing our builtin functions
this fix is avoiding such copies to happen

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 1]
```diff
@@ -106,14 +130,13 @@ internal TexlFunctionSet(IEnumerable<TexlFunctionSet> functionSets)
...
             }
 
-            foreach (var functionSet in functionSets)
-            {
-                functionSet._guard.VerifyNoWriters();
+            List<TexlFunctionSet> tfsList = functionSets.SelectMany(fs => fs.Sets).Where(fs => fs._count > 0).ToList();
 
-                if (functionSet._count > 0)
-                {
-                    Add(functionSet);
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:115]: >Add(functionSet) [](http://example.com/codeflow?start=20&length=16)

Key memory and performance issue

File: src/tests/.Net4.6.2/Microsoft.PowerFx.Connectors.Tests/Microsoft.PowerFx.Connectors.Tests.csproj [hunk 2]
```diff
@@ -14,8 +14,8 @@
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
     <Reference Include="System.Web" />
     
-    <PackageReference Include="Azure.Identity" Version="1.11.0" /> <!-- CVE-2024-29992 -->
-    <PackageReference Include="Microsoft.Identity.Client" Version="4.61.1" /> <!-- CVE-2024-27086 -->
+    <PackageReference Include="Azure.Identity" Version="1.11.4" />
```
Comment (MikeStall) [inline src/tests/.Net4.6.2/Microsoft.PowerFx.Connectors.Tests/Microsoft.PowerFx.Connectors.Tests.csproj:17]: >Identity [](http://example.com/codeflow?start=37&length=8)

is this a necessary  part of the hotfix? 

(especially considering this is a potential QFE candidate)

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs [hunk 3]
```diff
@@ -19,12 +19,12 @@ namespace Microsoft.PowerFx
     /// </summary>
     internal class ComposedReadOnlySymbolTable : ReadOnlySymbolTable, INameResolver, IGlobalSymbolNameResolver, IEnumStore
     {
-        private readonly IEnumerable<ReadOnlySymbolTable> _symbolTables;
+        private readonly List<ReadOnlySymbolTable> _symbolTables;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs:22]: >List [](http://example.com/codeflow?start=25&length=4)

do we need this change?

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs [hunk 4]
```diff
@@ -328,20 +328,22 @@ internal void RemoveFunction(TexlFunction function)
...
 
-            if (functions._count == 0)
+            if (functions.Count() == 0)
             {
                 return;
             }
-
-            _functions.Add(functions);
+
+            TexlFunctionSet funcsToAdd = makeCopy ? functions.Clone() : functions;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs:341]: >         TexlFunctionSet funcsToAdd = makeCopy ? functions.Clone() : functions; [](http://example.com/codeflow?start=3&length=79)

fwiw - I notice that if I switch this to always clone() (and ignore makeCopy), we still get same perf numbers in our benchmark.  

Is that expected?

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 5]
```diff
@@ -123,15 +150,23 @@ internal IEnumerable<TexlFunction> Functions
...
                     }
                 }
             }
-        }        
+        }
+
+        public void PreventRemoval()
+        {
+            _canremove = false;            
+        }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:169]: is there anything that calls this?

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 6]
```diff
@@ -28,45 +32,68 @@ internal class TexlFunctionSet
         // Count of functions
         internal int _count;
 
-        internal Dictionary<string, List<TexlFunction>>.KeyCollection FunctionNames => _functions.Keys;
+        internal bool _canremove;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:35]: I think this is a good idea. 
A concept of immutable sets is good and can help with optimizations.

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 7]
```diff
@@ -28,45 +32,68 @@ internal class TexlFunctionSet
         // Count of functions
         internal int _count;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:33]: with new Count() method - should this field now be private? 
Ie, should all callers go through Count() now?

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 8]
```diff
@@ -106,14 +134,13 @@ internal TexlFunctionSet(IEnumerable<TexlFunctionSet> functionSets)
...
 
-                if (functionSet._count > 0)
-                {
-                    Add(functionSet);
-                }
+            // keep local (empty) set for additional functions that could be added later
+            if (tfsList.Any())
+            {
+                _composed = true;
+                _texlFunctionSets = tfsList;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:143]: >        _texlFunctionSets = tfsList; [](http://example.com/codeflow?start=8&length=36)

Before - we took an (expensive) snapshot of the functionset.  So if the function set was modified later, we did not pick up those changes here. 

Now, we avoid the copy (win!) - but could that change the semantics?
Ie, do we need to verify that the incoming functionSets are not mutable?

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs (hunk 3 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs:22]: little perf improvement iirc but not stictly needed

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs (hunk 5 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:169]: right could be removed, dead code

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs (hunk 4 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs:341]: very possible as this method isn't used a lot IIRC

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs (hunk 8 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:143]: Here's a test demonstrating the change in behavior: #2478

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs (hunk 7 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:33]: yes we can make it private

File: src/tests/.Net4.6.2/Microsoft.PowerFx.Connectors.Tests/Microsoft.PowerFx.Connectors.Tests.csproj (hunk 2 above)
Comment (MikeStall) [reply src/tests/.Net4.6.2/Microsoft.PowerFx.Connectors.Tests/Microsoft.PowerFx.Connectors.Tests.csproj:17]: Latest main has this updates. Clearly, this should not be part of the hotfix; but let's confirm it goes away when you rebase on latest.

**Pull request closed without merging**